### PR TITLE
Fix is_iterator for container algorithms

### PR DESCRIPTION
**Pull request by dcbdan**
- Fix is_iterator so that SFINAE can work
- Call projection functions with invoke so that member functions
  can be passed as arguments.

### CODE REVIEW

File: hpx/parallel/algorithms/transform.hpp [hunk 1]
```diff
@@ -92,7 +93,8 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
                         [f, proj](reference t)
                         {
                             using hpx::util::get;
-                            get<1>(t) = f(proj(get<0>(t))); //-V573
+                            using hpx::util::invoke;
+                            get<1>(t) = f(invoke(proj, get<0>(t)));
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transform.hpp]: That should probably be `hpx::util::invoke`

File: hpx/parallel/algorithms/transform.hpp [hunk 2]
```diff
@@ -214,8 +216,9 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
             InIter2 first2,  OutIter dest, F && f,
             Proj1 && proj1, Proj2 && proj2)
         {
+            using hpx::util::invoke;
             while (first1 != last1)
-                *dest++ = f(proj1(*first1++), proj2(*first2++));
+                *dest++ = f(invoke(proj1, *first1++), invoke(proj2, *first2++));
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transform.hpp]: Same as above, should be `hpx::util::invoke`.

File: hpx/parallel/algorithms/transform.hpp [hunk 3]
```diff
@@ -420,8 +425,9 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
             InIter2 first2, InIter2 last2, OutIter dest, F && f,
             Proj1 && proj1, Proj2 && proj2)
         {
+            using hpx::util::invoke;
             while (first1 != last1 && first2 != last2)
-                *dest++ = f(proj1(*first1++), proj2(*first2++));
+                *dest++ = f(invoke(proj1, *first1++), invoke(proj2, *first2++));
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transform.hpp]: Same as above, should be `hpx::util::invoke`.

File: hpx/parallel/algorithms/transform.hpp [hunk 4]
```diff
@@ -476,7 +482,9 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
                         [f, proj1, proj2](reference t)
                         {
                             using hpx::util::get;
-                            get<2>(t) = f(proj1(get<0>(t)), proj2(get<1>(t))); //-V573
+                            using hpx::util::invoke;
+                            get<2>(t) = f(invoke(proj1, get<0>(t)),
+                                invoke(proj2, get<1>(t)));
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transform.hpp]: Same as above, should be `hpx::util::invoke`.

**Comment by hkaiser** [comment]
lgtm, thanks!

**Comment by AntonBikineev** [comment]
@hkaiser: just out of curiosity, what's benefit of using explicit quailification of hpx::util::invoke instead of using declarations?

**Comment by hkaiser** [comment]
> @hkaiser: just out of curiosity, what's benefit of using explicit quailification of hpx::util::invoke instead of using declarations in this case?

This allows for things like this:
```
struct A { 
    int foo() { return 42; }
};

std::vector<A> v(43);

parallel::for_each(
    par, begin(v), end(v)
    [](int i) { cout << i; },
    &A::foo);
```

**Comment by K-ballo** [comment]
> what's benefit of using explicit quailification of hpx::util::invoke instead of using declarations in this case?

@AntonBikineev It avoids ADL havoc, qualified vs unqualified is always about avoiding ADL. For instance, when `std::` arguments are used both `hpx::util::invoke` and `std::invoke` are viable matches and the unqualified call is ambiguous. Generic libraries should always do qualified calls unless they explicitly want a call to reach a customization point.

**Comment by AntonBikineev** [comment]
@K-ballo: thanks for the point. What about this:
using hpx::util::invoke;
(invoke)(something);
/*joke*/

**Pull request merged**